### ISSUE TITLE
DOM images textures break in Firefox

### ISSUE THREAD

**Issue by shepelevstas** [joakin/elm-canvas#24]
I have an app that gets a list of files and sends them as `List D.Value` to a port, on js side I do this:
File -> Image -> Canvas -> downscale -> Canvas -> Image -> send it back to elm through port.
I send Images back one by one as soon as any of them is ready. Then in elm I decode them with `Canvas.Texture.fromDomImage` and save in a list in model, then draw them in view with `List.map`.
Often that breaks elm app in Firefox. Sometimes works ok. With the same images. When it breaks - it is always the same error, I made screenshots.
![Capture1](https://user-images.githubusercontent.com/14919836/80232053-74b14f80-867e-11ea-83f9-28e064a54d36.PNG)
![Capture2](https://user-images.githubusercontent.com/14919836/80232060-78dd6d00-867e-11ea-81de-51d07749ea51.PNG)
I've updated Firefox to the latest version. This does not happen in Chrome or Safari.
I really like this package! I'd love to help anyway I can!
The app I am working on is here:
[https://github.com/shepelevstas/elmphoto.git](https://github.com/shepelevstas/elmphoto.git)

**Comment by joakin** [comment]
The Elm and JS library should be deterministic. You say:

> Often that breaks elm app in Firefox. Sometimes works ok.

That doesn't sound deterministic at all. 

Are your images sizes valid? Check on them before sending them to elm, the width and height. Maybe it is this: https://bugzilla.mozilla.org/show_bug.cgi?id=1497275

Maybe this helps?

* https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/drawImage#Exceptions_thrown
* https://stackoverflow.com/questions/19338032/canvas-indexsizeerror-index-or-size-is-negative-or-greater-than-the-allowed-a

**Comment by shepelevstas** [comment]
Seems to be Firefox problem=) Thanks for the links! I'll try solutions from the last link!

**Comment by shepelevstas** [comment]
Looks like DOM Image fails to be decoded at first try for every new image. I load two files, three times the first and then three times the second. For each file the first load results in TImage with height and width equal 0. Then drawImage command fails in Firefox.
![Capture](https://user-images.githubusercontent.com/14919836/80799505-36151b00-8bd1-11ea-98a0-a4015dde4184.PNG)

**Comment by shepelevstas** [comment]
Accedentally closed the issue. I'll try to create Ellie project!

**Comment by joakin** [comment]
Seems like  you may have problems in the JS.

Just to be clear, `fromDomImage` requires a image instance from the JS `Image` constructor, and defined `width` and `height` properties because there are necessary for `Texture.dimensions`.

**Comment by shepelevstas** [comment]
Yes, you are right! While creating Ellie example, I've noticed that in one of the functions that downscale the image - `image` was returned right after `image.src = canvas.toDataURL()`. I refactored it into a Promise and the error is gone! =D
I wonder how come Chrome and Safari never stumble on it =))
Thanks you for your help! You suggested creating an Ellie example in another issue, and it helped me to find the error!
And thank you for this great elm package! Works great and easy to use =D
